File: ships-preview/src/main/terminal/executor.ts
/**
 * Terminal Executor Module
 * 
 * Executes validated commands using child_process.spawn.
 * Handles streaming output, timeouts, and cleanup.
 */

import { spawn, ChildProcess } from 'child_process';
import { platform } from 'os';
import type { CommandRequest, CommandResult, CommandStreamEvent } from './types';
import { validateCommand, getAllowedCommandConfig } from './security';

/** Active processes for cleanup */
const activeProcesses: Map<string, ChildProcess> = new Map();

/**
 * Generate a unique process ID.
 */
function generateProcessId(): string {
  return `proc_${Date.now()}_${Math.random().toString(36).slice(2, 8)}`;
}

/**
 * Get the shell configuration for the current platform.
 */
function getShellConfig(): { shell: string; shellFlag: string } {
  if (platform() === 'win32') {
    return { shell: 'cmd.exe', shellFlag: '/c' };
  }
  return { shell: '/bin/sh', shellFlag: '-c' };
}

/**
 * Execute a command and return the result.
 * This is the main execution function.
 */
export async function executeCommand(request: CommandRequest): Promise<CommandResult> {
  const startTime = Date.now();
  const processId = generateProcessId();
  
  // Validate command first
  const validation = validateCommand(request.command, request.cwd);
  if (!validation.isValid) {
    return {
      success: false,
      exitCode: null,
      stdout: '',
      stderr: validation.error || 'Command validation failed',
      error: validation.error,
      durationMs: Date.now() - startTime,
    };
  }

  // Get timeout from allowed command config or request
  const cmdConfig = getAllowedCommandConfig(request.command);
  const timeout = request.timeout ?? cmdConfig?.maxTimeout ?? 60000;

  const { shell, shellFlag } = getShellConfig();
  
  return new Promise((resolve) => {
    let stdout = '';
    let stderr = '';
    let timedOut = false;
    let resolved = false;

    const finish = (exitCode: number | null, error?: string) => {
      if (resolved) return;
      resolved = true;
      
      // Cleanup
      activeProcesses.delete(processId);
      
      resolve({
        success: exitCode === 0,
        exitCode,
        stdout,
        stderr,
        error: error || (timedOut ? 'Command timed out' : undefined),
        timedOut,
        durationMs: Date.now() - startTime,
      });
    };

    try {
      const child = spawn(shell, [shellFlag, validation.sanitizedCommand!], {
        cwd: request.cwd,
        env: {
          ...process.env,
          // Ensure npm doesn't prompt for input
          npm_config_yes: 'true',
          CI: 'true',
        },
        windowsHide: true,
      });

      activeProcesses.set(processId, child);

      // Set timeout
      const timeoutHandle = setTimeout(() => {
        timedOut = true;
        child.kill('SIGTERM');
        
        // Force kill after 5 seconds if still running
        setTimeout(() => {
          if (!resolved) {
            child.kill('SIGKILL');
          }
        }, 5000);
      }, timeout);

      child.stdout?.on('data', (data: Buffer) => {
        stdout += data.toString();
      });

      child.stderr?.on('data', (data: Buffer) => {
        stderr += data.toString();
      });

      child.on('error', (error: Error) => {
        clearTimeout(timeoutHandle);
        finish(null, error.message);
      });

      child.on('close', (code: number | null) => {
        clearTimeout(timeoutHandle);
        finish(code);
      });

    } catch (error: any) {
      finish(null, error.message);
    }
  });
}

/**
 * Execute a command with streaming output.
 * Calls the callback for each output event.
 */
export async function executeCommandWithStream(
  request: CommandRequest,
  onEvent: (event: CommandStreamEvent) => void
): Promise<CommandResult> {
  const startTime = Date.now();
  const processId = generateProcessId();
  
  // Validate command first
  const validation = validateCommand(request.command, request.cwd);
  if (!validation.isValid) {
    const error = validation.error || 'Command validation failed';
    onEvent({ type: 'error', data: error, timestamp: Date.now() });
    return {
      success: false,
      exitCode: null,
      stdout: '',
      stderr: error,
      error,
      durationMs: Date.now() - startTime,
    };
  }

  // Get timeout from allowed command config or request
  const cmdConfig = getAllowedCommandConfig(request.command);
  const timeout = request.timeout ?? cmdConfig?.maxTimeout ?? 60000;

  const { shell, shellFlag } = getShellConfig();
  
  return new Promise((resolve) => {
    let stdout = '';
    let stderr = '';
    let timedOut = false;
    let resolved = false;

    const finish = (exitCode: number | null, error?: string) => {
      if (resolved) return;
      resolved = true;
      
      activeProcesses.delete(processId);
      
      onEvent({ 
        type: 'exit', 
        data: String(exitCode ?? 'null'), 
        timestamp: Date.now() 
      });
      
      resolve({
        success: exitCode === 0,
        exitCode,
        stdout,
        stderr,
        error: error || (timedOut ? 'Command timed out' : undefined),
        timedOut,
        durationMs: Date.now() - startTime,
      });
    };

    try {
      const child = spawn(shell, [shellFlag, validation.sanitizedCommand!], {
        cwd: request.cwd,
        env: {
          ...process.env,
          npm_config_yes: 'true',
          CI: 'true',
        },
        windowsHide: true,
      });

      activeProcesses.set(processId, child);

      const timeoutHandle = setTimeout(() => {
        timedOut = true;
        onEvent({ type: 'error', data: 'Command timed out', timestamp: Date.now() });
        child.kill('SIGTERM');
        setTimeout(() => {
          if (!resolved) child.kill('SIGKILL');
        }, 5000);
      }, timeout);

      child.stdout?.on('data', (data: Buffer) => {
        const text = data.toString();
        stdout += text;
        onEvent({ type: 'stdout', data: text, timestamp: Date.now() });
      });

      child.stderr?.on('data', (data: Buffer) => {
        const text = data.toString();
        stderr += text;
        onEvent({ type: 'stderr', data: text, timestamp: Date.now() });
      });

      child.on('error', (error: Error) => {
        clearTimeout(timeoutHandle);
        onEvent({ type: 'error', data: error.message, timestamp: Date.now() });
        finish(null, error.message);
      });

      child.on('close', (code: number | null) => {
        clearTimeout(timeoutHandle);
        finish(code);
      });

    } catch (error: any) {
      onEvent({ type: 'error', data: error.message, timestamp: Date.now() });
      finish(null, error.message);
    }
  });
}

/**
 * Kill all active processes.
 * Called during app shutdown.
 */
export function killAllProcesses(): void {
  for (const [_id, proc] of activeProcesses) {
    try {
      proc.kill('SIGTERM');
    } catch {
      // Process may already be dead
    }
  }
  activeProcesses.clear();
}

/**
 * Get count of active processes.
 */
export function getActiveProcessCount(): number {
  return activeProcesses.size;
}

File: ships-backend/app/terminal/executor.py
"""
Terminal Executor Module

Executes validated commands using subprocess.
Handles timeouts and output streaming.
"""

import subprocess
import time
import platform
import logging
from typing import Optional, Callable
from pathlib import Path

from .models import CommandRequest, CommandResult
from .security import validate_command, get_allowed_command_config

logger = logging.getLogger("ships.terminal")


def get_shell_config() -> tuple[str, list[str]]:
    """Get shell configuration for the current platform."""
    if platform.system() == "Windows":
        return "cmd.exe", ["/c"]
    return "/bin/sh", ["-c"]


async def execute_command(request: CommandRequest) -> CommandResult:
    """
    Execute a command and return the result.
    
    This is the main async execution function for agent tools.
    """
    start_time = time.time()
    
    # Validate first
    is_valid, error, sanitized = validate_command(request.command, request.cwd)
    if not is_valid:
        logger.warning(f"[EXECUTOR] ❌ Validation failed: {error}")
        return CommandResult(
            success=False,
            error=error,
            duration_ms=int((time.time() - start_time) * 1000)
        )
    
    # Get timeout from config
    cmd_config = get_allowed_command_config(request.command)
    timeout = request.timeout or (cmd_config.max_timeout if cmd_config else 60)
    
    shell, shell_args = get_shell_config()
    
    try:
        logger.info(f"[EXECUTOR] 🚀 Running: {sanitized}")
        
        # Run the command
        result = subprocess.run(
            [shell] + shell_args + [sanitized],
            cwd=request.cwd,
            capture_output=True,
            timeout=timeout,
            text=True,
            env={
                **subprocess.run.__globals__.get("os", __import__("os")).environ,
                "CI": "true",
                "npm_config_yes": "true",
            }
        )
        
        duration = int((time.time() - start_time) * 1000)
        
        logger.info(f"[EXECUTOR] {'✅' if result.returncode == 0 else '❌'} Exit code: {result.returncode} in {duration}ms")
        
        return CommandResult(
            success=result.returncode == 0,
            exit_code=result.returncode,
            stdout=result.stdout[:10000] if result.stdout else "",  # Limit output size
            stderr=result.stderr[:5000] if result.stderr else "",
            duration_ms=duration
        )
        
    except subprocess.TimeoutExpired as e:
        duration = int((time.time() - start_time) * 1000)
        logger.warning(f"[EXECUTOR] ⏰ Command timed out after {timeout}s")
        return CommandResult(
            success=False,
            error=f"Command timed out after {timeout} seconds",
            stdout=e.stdout[:5000] if e.stdout else "",
            stderr=e.stderr[:2500] if e.stderr else "",
            timed_out=True,
            duration_ms=duration
        )
        
    except Exception as e:
        duration = int((time.time() - start_time) * 1000)
        logger.error(f"[EXECUTOR] ❌ Execution error: {e}")
        return CommandResult(
            success=False,
            error=str(e),
            duration_ms=duration
        )


def execute_command_sync(request: CommandRequest) -> CommandResult:
    """
    Synchronous version for non-async contexts.
    """
    import asyncio
    return asyncio.get_event_loop().run_until_complete(execute_command(request))
